pkg/preparer/setup.go

```diff
@@ -650,6 +652,71 @@ func getArtifactRegistry(preparerConfig *PreparerConfig) (artifact.Registry, err
 	return artifact.NewRegistry(url, fetcher, osversion.DefaultDetector), nil
 }
 
+func (p *Preparer) BuildRealityAtLaunch() error {
+	// check if reality tree does not exist, if true then reimage occurred and may need
+	// to rebuild reality tree for orphaned pods
+	// a pod is orphaned if they exist on disk but not in intent
+	sub := p.Logger.SubLogger(nil)
+	realityResults, _, err := p.store.ListPods(consul.REALITY_TREE, p.node)
+	if err != nil {
+		sub.WithError(err).Errorln("Could not check reality: %s", err)
+		return err
+	} else if len(realityResults) == 0 {
+		// get set of pods in intent
+		intentResults, _, err := p.store.ListPods(consul.INTENT_TREE, p.node)
+		if err != nil {
+			sub.WithError(err).Errorln("Could not check intent: %s", err)
+			return err
+		}
+		intentMap := map[string]bool{}
+		for _, result := range intentResults {
+			intentMap[result.Manifest.ID().String()] = true
+		}
+
+		// get set of pods installed on machine
+		manifestFilePaths, err := filepath.Glob("/data/pods/*/current_manifest.yaml")
+		if err != nil {
+			sub.WithError(err).Errorln("Could not check filepaths of pods installed on disk: %s", err)
+			return err
+		}
+		for _, fp := range manifestFilePaths {
+			podIDStr := strings.SplitN(fp, "/", 5)[3]
+			var podUUID uuid.UUID
+			if len(podIDStr) > types.PodUUIDLength {
+				podUUID = uuid.Parse(podIDStr[len(podIDStr)-types.PodUUIDLength:])
```

Comment:
probably this should be extracted to a function since it's doing the same thing as `PodWithPodHome`